rs/sns/swap/src/swap.rs

```diff
@@ -910,14 +911,20 @@ impl Swap {
                         total_participant_icp_e8s,
                     );
 
+                    #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
                     let Some(parsed_principal) =
-                        string_to_principal(&cf_participant.hotkey_principal)
+                    // TODO(NNS1-3198): Simplify once hotkey_principal is removed
+                    cf_participant
+                        .controller
+                        .or(string_to_principal(&cf_participant.hotkey_principal))
```

Comment:
Counterpoint: by using `.or` systematically, we don't ever have to do some dataflow analysis to make sure we've covered every case. In that sense we can just be monkeys who mechanically substitute `hotkey_principal` with `controller.or(hotkey_principal)` and know that the outcome will be code that behaves correct. If this were going to live for longer, I would agree with making the code clearer by doing what you suggested, but since this will be cleaned up by the ticket linked in the comment I think the benefit is lower than the cost of making sure we don't miss a case where we should have normalized and didn't.

If we wanted to go the normalization route, I would suggest something that I think would be even better and that we should do anyway: separate our API types from the types we use for internal logic, have the API type only have a non-optional `controller` field, and do this normalization at the point of conversion. Then any code that uses the internal-logic type would be guaranteed at the type level to have an unambiguous `controller`.